pages/deployments/oidc/oidc_with_aws.md

```diff
@@ -0,0 +1,187 @@
+
+- **Provider URL**: `https://agent.buildkite.com`
+
+- **Audience**: `sts.amazonaws.com`
+
+## Step 2: Create a new (or update an existing) IAM role to use with your pipelines
+
+Creating new or updating existing IAM roles is conducted through your AWS account.
+
+Learn more about how to do this in the [Creating a role using custom trust policies (console)](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-custom.html) page of the AWS documentation.
+
+As part of this process:
+
+1. Choose the **Custom trust policy** role type.
+
+1. Copy the following example trust policy in the following JSON code block and paste it into a code editor:
+
+    ```json
+    {
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Effect": "Allow",
+                "Principal": {
+                    "Federated": "arn\:aws\:iam:\:AWS_ACCOUNT_ID\:oidc-provider/agent.buildkite.com"
+                },
+                "Action": "sts:AssumeRoleWithWebIdentity",
+                "Condition": {
+                    "StringEquals": {
+                        "agent.buildkite.com:aud": "sts.amazonaws.com"
+                    },
+                    "StringLike": {
+                        "agent.buildkite.com:sub": "organization\:ORGANIZATION_SLUG\:pipeline:PIPELINE_SLUG\:ref\:REF\:commit\:BUILD_COMMIT\:step\:STEP_KEY"
+                    },
+                    "IpAddress": {
+                        "aws:SourceIp": [
+                            "AGENT_PUBLIC_IP_ONE",
+                            "AGENT_PUBLIC_IP_TWO"
+                        ]
+                    }
+
+                }
+            }
+        ]
+    }
+    ```
+
+1. Modify the following sections of the pasted code snippet accordingly:
+    * `AWS_ACCOUNT_ID` (in `Principal`) is your AWS account ID.
+    * next item
+
+1. In the **Custom trust policy** section,
+
+1. Specify an appropriate **Role name**, for example, `compute-ssm-oidc-example`.
+
+Set the Principal to be federated via the IAM OIDC provider we just created, and add some conditions against using this IAM role for additional security using the format of the OIDC token subject provided by Buildkite.
+
+This allows us to limit use of this IAM role to named Organizations, Pipelines, and optionally branches, commits and steps. Format is
+ `organization:ORGANIZATION_SLUG:pipeline:PIPELINE_SLUG:ref:REF:commit:BUILD_COMMIT:step:STEP_KEY`
+
+You can use a wildcard `*` to replace any subjects you don’t want to set limits on.
+
+You can also add multiple token subjects as a list if you want to use the same IAM role for multiple pipelines, or if you want a variety of branches, commits and steps enabled.
+
+If you have a public IP address associated with your Buildkite Agent's, you can also set a condition on the source IP address and further restrict access to your IAM role.
```

Comment:
Does this relate to the:

```
                    "IpAddress": {
                        "aws:SourceIp": [
                            "AGENT_PUBLIC_IP_ONE",
                            "AGENT_PUBLIC_IP_TWO"
                        ]
                    }
```

part of the trust policy (JSON) code snippet?

Reply:
Yes that's right. We can say that only token exchange requests coming from those IP addresses are valid, and it'll ignore requests from anywhere else.

Customers don't have to assign IP addresses to their agents, and in the new hosted agents they _cannot_ assign IP addresses to their agents, which is why it's being talked about as an aside.

It's a really useful option for folks to use though, and we use it extensively internally.

Reply:
Useful to know! Thanks, I'll put some of this key info into the first cut of these docs.